### DISCUSSION TITLE
YOLO Pose Estimation - Keypoints are flipped

### DISCUSSION THREAD

**Discussion by groundfzero** [ultralytics/ultralytics#22803] [Solutions]
I trained a YOLO v8 model (using approximately 4000 images) to detect keypoints around the object. My object is cuboid shaped but it has features on every faces that are distinct. When I ran the prediction, 5-8% of the time keypoints are flipped (appears as if they are rotated by 180 degrees about the Y-axis of the 2D image plane) and my rotational error for the object goes to 180 degrees as well. I am curious if this has to do with my training, the shape of the object or something else. Any help is appreciated.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @groundfzero, thanks for opening this discussion and for using Ultralytics for pose estimation 🚀  
This is an automated response to help you get support as quickly as possible — an Ultralytics engineer will review your post and assist you further soon 🙂

Since your question involves YOLO pose estimation and potential keypoint flipping, it may be either:
- a 🐛 Bug Report (unexpected model behavior), or  
- a custom training ❓ Question (data, labels, or training configuration related)

Please review the points below so we have enough detail to help you effectively 👇

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. This is especially important for intermittent issues like the keypoints flipping you described.

If this is a custom training ❓ Question, please provide as much information as possible, including:
- example training images of your cuboid object (with keypoints visible)
- how your keypoints are defined and labeled (order and semantics)
- training logs and key hyperparameters
- any post-processing or pose-to-rotation conversion you are doing  
and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

You can also check the [Docs](https://docs.ultralytics.com/) for additional details on pose models and see many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples that might help clarify configuration or inference details.

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- For longer-form help, visit [Discourse](https://community.ultralytics.com/)  
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community

## Upgrade

Before further debugging, please upgrade to the latest `ultralytics` package to ensure you are not seeing an issue already fixed in a newer release. Use a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) and install all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml):

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
This kind of 180° keypoint swap on a cuboid is almost always down to data/label ambiguity rather than a runtime bug: if some faces of the object look similar, the model effectively has two nearly equivalent solutions and will sometimes choose the “mirrored” one. To narrow it down, please double‑check that your keypoint order and any `flip_idx` mapping in your dataset YAML are correct (these are used when applying random flips during training, see the augmentation reference in the [data/augment docs](https://docs.ultralytics.com/reference/data/augment/)), and try a short ablation run with horizontal flip disabled to see if it reduces the failures, for example: 

```bash
yolo pose train model=your-model.pt data=your-data.yaml fliplr=0.0
``` 

If you can share a couple of representative images with ground‑truth keypoints overlaid plus the corresponding bad predictions, we can comment more specifically on whether it looks like inherent symmetry vs. mislabeled or conflicting views.
